File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if(preorder.empty()) return nullptr;

        TreeNode* root = new TreeNode(preorder[0]);
        
        int index_of_root = -1;
        for(int i=0; i<preorder.size(); i++){
            if(inorder[i]==preorder[0]){
                index_of_root = i;
                break;
            }
        }

        vector<int> left_preorder(preorder.begin()+1, preorder.begin()+index_of_root+1);
        vector<int> left_inorder(inorder.begin(), inorder.begin()+index_of_root);
        
        vector<int> right_preorder(preorder.begin()+index_of_root+1, preorder.end());
        vector<int> right_inorder(inorder.begin()+index_of_root+1, inorder.end());

        TreeNode* left = buildTree(left_preorder, left_inorder);
        TreeNode* right = buildTree(right_preorder, right_inorder);

        root->left = left;
        root->right = right;

        return root;
    }
};

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode helper(int ps, int pe, int[] preorder,int is, int ie, int[] inorder){
        if(is>ie || ps>pe ){
            return null;
        }
        TreeNode n = new TreeNode(preorder[ps]);
        int count = is;
        while(inorder[count]!=preorder[ps]){
            count++;
        }
        
        n.left=helper(ps+1,ps+count-is,preorder,is,count-1,inorder);
        n.right=helper(ps+count-is+1,pe,preorder,count+1,ie,inorder);
        return n;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return helper(0,preorder.length-1,preorder,0,inorder.length-1,inorder);
    }
}